### PR TITLE
[MAINT]  Examples output all their results in a single main directory

### LINKED ISSUE #4209: Ensure all examples output to their files to a git ignored directory
Not a big issue but I am slowly getting irritated of reverting commits because I committed 50 niftis to the repo.

This may also simplify our .gitignore to avoid things like:

```
# examples
examples/**/*.pdf
examples/**/results/
examples/results/
results/
examples/*.nii
examples/*.nii.gz
examples/*.png
examples/04_glm_first_level/derivatives
```

### LINKED ISSUE #4228: title of figure overlaps with progress bar in `plot_first_level_details` example
minor issue

see in 
https://nilearn.github.io/dev/auto_examples/04_glm_first_level/plot_first_level_details.html

image

note also that the name of the different conditions overlap

I think we should have:

"left - right button press"
"audio - visual"
"computation - sentences"
"horizontal-vertical"

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Closes #4228 
- Closes #4209 

Changes proposed in this pull request:

- make sure that all example output to "nilearn/results/example_name"

Also in this PR:
- use `Path` instead of `os` in examples
- fix some figures so that title would not overlap with color bar or be too wide for the figure
- increase stack level of the warning of some private functions
- added some checks in all examples to make sure that matplotlib is installed
  - when an example use pyplot AND NO some nilearn.plotting function, then there a try import block at the top
  - when an example use pyplot AND some nilearn.plotting function, the import nilearn.plotting was hoisted at the top of the example
  - when this was not possible (because isort putting the import matplotlib.pyplot before the nilearn.plotting imports) then a try import block was added

### CODE REVIEW

File: examples/01_plotting/plot_demo_glass_brain_extensive.py [hunk 1]
```diff
@@ -93,7 +93,7 @@
 # colorbar will not be centered around zero.
 
 # only plot positive values
-plotting.plot_glass_brain(
+plot_glass_brain(
```
Comment (Remi-Gau) [inline examples/01_plotting/plot_demo_glass_brain_extensive.py:96]: to make it more consistent with the rest of the example

File: examples/03_connectivity/plot_multi_subject_connectome.py [hunk 2]
```diff
@@ -101,11 +101,8 @@ def plot_matrices(cov, prec, title, labels):
 gsc = GroupSparseCovarianceCV(verbose=2)
 gsc.fit(subject_time_series)
 
-try:
-    from sklearn.covariance import GraphicalLassoCV
-except ImportError:
-    # for Scitkit-Learn < v0.20.0
-    from sklearn.covariance import GraphLassoCV as GraphicalLassoCV
```
Comment (Remi-Gau) [inline examples/03_connectivity/plot_multi_subject_connectome.py:108]: am I correct that this is not needed anymore given the scikit learn requirements we have?

File: examples/05_glm_second_level/plot_oasis.py [hunk 3]
```diff
@@ -32,6 +32,12 @@
     - Gael Varoquaux, Apr 2014
 
 """
+# %%
+try:
+    import matplotlib.pyplot as plt
+except ImportError:
+    raise RuntimeError("This script needs the matplotlib library")
+
```
Comment (Remi-Gau) [inline examples/05_glm_second_level/plot_oasis.py:40]: some examples have this where as others do not: 

I fell that if they are plotting things they should all have it (or none should have it)

File: examples/auditory_block_paradigm.csv
Comment (Remi-Gau) [inline examples/auditory_block_paradigm.csv:1]: could not identify what this file is for so just removing it to see if anything fails

File: examples/03_connectivity/plot_multi_subject_connectome.py (hunk 2 above)
Comment (bthirion) [reply examples/03_connectivity/plot_multi_subject_connectome.py:108]: indeed.

File: examples/04_glm_first_level/plot_first_level_details.py [hunk 4]
```diff
@@ -198,7 +198,7 @@ def plot_contrast(first_level_model):
     design_matrix = first_level_model.design_matrices_[0]
     # Call the contrast specification within the function
     contrasts = make_localizer_contrasts(design_matrix)
-    plt.figure(figsize=(11, 3))
+    plt.figure(figsize=(20, 4))
```
Comment (bthirion) [inline examples/04_glm_first_level/plot_first_level_details.py:201]: Why this change ?

File: examples/05_glm_second_level/plot_oasis.py (hunk 3 above)
Comment (bthirion) [reply examples/05_glm_second_level/plot_oasis.py:40]: Indeed

Comment (bthirion) [reply examples/05_glm_second_level/plot_oasis.py:40]: They should all have it.

**Review by bthirion — COMMENTED** [review]
The changes look good so far.

File: examples/04_glm_first_level/plot_first_level_details.py (hunk 4 above)
Comment (Remi-Gau) [reply examples/04_glm_first_level/plot_first_level_details.py:201]: to avoid the isssue of #4228 where subplot title overlap with each other and with the color bar

File: examples/05_glm_second_level/plot_oasis.py (hunk 3 above)
Comment (Remi-Gau) [reply examples/05_glm_second_level/plot_oasis.py:40]: OK will update
- [ ] add to all examples

Comment (Remi-Gau) [reply examples/05_glm_second_level/plot_oasis.py]: OK what I have done is to add this try / except block in all examples that use `pyplot` but NOT `nilearn.plotting` because if they import from the latter they will anyway get a warning if matplotlib is not installed

Otherwise I was ending up adding this block in literally all the examples.

**Review by bthirion — APPROVED** [review]
LGTM, thx !

File: examples/05_glm_second_level/plot_oasis.py [hunk 5]
```diff
@@ -90,13 +89,16 @@
     columns=["age", "sex", "intercept"],
 )
 
+from matplotlib import pyplot as plt
```
Comment (ymzayek) [inline examples/05_glm_second_level/plot_oasis.py:92]: don't we also need to catch the ImportError in this example?

Comment (ymzayek) [reply examples/05_glm_second_level/plot_oasis.py:92]: Or move the `from nilearn import plotting` above this import

**Review by ymzayek — COMMENTED** [review]
Just one thing, otherwise LGTM!

File: examples/05_glm_second_level/plot_oasis.py (hunk 5 above)
Comment (Remi-Gau) [reply examples/05_glm_second_level/plot_oasis.py:92]: ha yeah that makes sense: will do that for the examples that import pyplot before nilearn plotting

Comment (Remi-Gau) [reply examples/05_glm_second_level/plot_oasis.py:92]: OK the logic got a bit more complicated:
- when an example use pyplot, then there a try import block at the top
- when an example use pyplot AND some nilearn.plotting function I hoisted the import nilearn.plotting at the top of the example
- when this was not possible (because isort putting the import matplotlib.pyplot before the nilearn.plotting imports) then I added a try import block

**Review by bthirion — APPROVED** [review]
All right.

File: examples/07_advanced/plot_beta_series.py [hunk 6]
```diff
@@ -58,11 +58,14 @@
...
+    import matplotlib.pyplot as plt
+except ImportError:
+    raise RuntimeError("This script needs the matplotlib library")
+
 # sphinx_gallery_thumbnail_number = -2
 
 # %%
-import matplotlib.pyplot as plt
-
 from nilearn import image, plotting
```
Comment (ymzayek) [inline examples/07_advanced/plot_beta_series.py:69]: here can't you also just switch the positions of importing nilearn.plotting and pyplot instead of raising the import error?

File: examples/08_experimental/plot_surface_image_and_maskers.py [hunk 7]
```diff
@@ -14,13 +14,15 @@
...
+except ImportError:
+    raise RuntimeError("This script needs the matplotlib library")
 
 # %%
 
 from typing import Optional, Sequence
 
-from matplotlib import pyplot as plt
-
 from nilearn import plotting
```
Comment (ymzayek) [inline examples/08_experimental/plot_surface_image_and_maskers.py:26]: here also

File: examples/01_plotting/plot_colormaps.py [hunk 8]
```diff
@@ -7,18 +7,19 @@
...
+# ---------------
+try:
+    import matplotlib.pyplot as plt
+except ImportError:
+    raise RuntimeError("This script needs the matplotlib library")
+
 import numpy as np
 
 from nilearn.plotting import show
 from nilearn.plotting.cm import _cmap_d as nilearn_cmaps
```
Comment (ymzayek) [inline examples/01_plotting/plot_colormaps.py:21]: and here

File: examples/01_plotting/plot_surface_projection_strategies.py [hunk 9]
```diff
@@ -21,11 +21,14 @@
...
+    import matplotlib.pyplot as plt
+except ImportError:
+    raise RuntimeError("This script needs the matplotlib library")
 
 # %%
 import matplotlib
 import numpy as np
-from matplotlib import pyplot as plt
 
 from nilearn.plotting import show
```
Comment (ymzayek) [inline examples/01_plotting/plot_surface_projection_strategies.py:33]: also here

File: examples/02_decoding/plot_haxby_stimuli.py [hunk 10]
```diff
@@ -6,10 +6,12 @@
...
+try:
+    import matplotlib.pyplot as plt
+except ImportError:
+    raise RuntimeError("This script needs the matplotlib library")
 
 # %%
-import matplotlib.pyplot as plt
-
 from nilearn import datasets
 from nilearn.plotting import show
```
Comment (ymzayek) [inline examples/02_decoding/plot_haxby_stimuli.py:16]: same here

File: examples/07_advanced/plot_beta_series.py (hunk 6 above)
Comment (Remi-Gau) [reply examples/07_advanced/plot_beta_series.py:69]: If I do this then we get an isort error:


```
git diff

examples/07_advanced/plot_beta_series.py

────────────────────────┐
67: except ImportError: │
────────────────────────┘
│ 67 │                                                              │ 67 │
│ 68 │# %%                                                          │ 68 │# %%
│ 69 │from nilearn import image, plotting                           │ 69 │from nilearn import image, plotting
│    │                                                              │ 70 │import matplotlib.pyplot as plt
│ 70 │                                                              │ 71 │
│ 71 │# %%                                                          │ 72 │# %%
│ 72 │# Prepare data and analysis parameters                        │ 73 │# Prepare data and analysis parameters

(base) ╭─remi at remi-XPS-15-9570 in ~/github/nilearn/nilearn on output_examples✘✘✘ 24-01-22 - 9:48:39
╰─(base) ⠠⠵ pre-commit run -a
check python ast.........................................................Passed
check for case conflicts.................................................Passed
check json...............................................................Passed
check for merge conflicts................................................Passed
check toml...............................................................Passed
check yaml...............................................................Passed
fix end of files.........................................................Passed
mixed line ending........................................................Passed
trim trailing whitespace.................................................Passed
isort....................................................................Failed
- hook id: isort
- files were modified by this hook

Fixing /home/remi/github/nilearn/nilearn/examples/07_advanced/plot_beta_series.py
```

Comment (Remi-Gau) [reply examples/07_advanced/plot_beta_series.py:69]: Could try to move this import closer to where it is used though

Comment (ymzayek) [reply examples/07_advanced/plot_beta_series.py:69]: No worries it's very minor so can just keep it as it is. I did't think about isort...

**Review by ymzayek — APPROVED** [review]
Other than the import order (which you can leave as is) I think this one is good to go
